src/Rings/groebner.jl

```diff
@@ -1221,3 +1243,124 @@ function _compute_groebner_basis_using_fglm(I::MPolyIdeal,
 	dim(I) == 0 || error("Dimension of ideal must be zero.")
 	I.gb[destination_ordering] = _fglm(G, destination_ordering)
 end
+
+
+@doc Markdown.doc"""
+    groebner_basis_hilbert_driven(I::MPolyIdeal{P};
+                                  ordering::MonomialOrdering,
+                                  complete_reduction::Bool = false) where {P <: MPolyElem_dec}
+
+
+Compute a Gröbner basis of `I` with respect to `ordering` using a
+Hilbert Series driven method as follows: If a Gröbner basis for `I` is
+present, compute the Hilbert series of `I` and use it to optimize the
+Gröbner basis computation for `I` w.r.t. `ordering`. If no Gröbner
+basis for `I` is present compute the Hilbert series for `I` if the
+base field of `I` has positive characteristic, otherwise compute the
+Hilbert series for `I` modulo a randomly chosen prime. Use the
+resulting Hilbert series to optimize the Gröbner basis computation for
+`I` w.r.t. `ordering`.
+
+`I` must be given by generators homogeneous w.r.t. `weights`.
+
+# Examples
+```jldoctest
+julia> R, (x, y, z) = grade(PolynomialRing(QQ, ["x", "y", "z"])[1]);
+
+julia> I = ideal(R, [x^2 + y*z, x*y - y*z]);
+
+julia> groebner_basis_hilbert_driven(I, ordering = deglex(R), complete_reduction = true)
+Gröbner basis with elements
+1 -> x*y - y*z
+2 -> x^2 + y*z
+3 -> y^2*z + y*z^2
+with respect to the ordering
+deglex([x, y, z])
+```
+"""
+
+function groebner_basis_hilbert_driven(I::MPolyIdeal{P};
+                                       ordering::MonomialOrdering,
+                                       complete_reduction::Bool = false) where {P <: MPolyElem_dec}
+  
+  all(p -> is_homogeneous(p), gens(I)) || error("I must be given by generators homogeneous with respect to its underlying ring.")
+	isa(coefficient_ring(base_ring(I)), AbstractAlgebra.Field) || error("The underlying coefficient ring of I must be a field.")
+  is_global(ordering) || error("Destination ordering must be global.")
+	haskey(I.gb, ordering) && return I.gb[ordering]
+  if isempty(I.gb) && iszero(characteristic(base_ring(I)))  
+    while true
+      r = rand(2^15:2^16)
+      p = Hecke.next_prime(r)
```

Comment:
Just wondering, but: is there any particular reason to pick a random `r` each time (and thus potentially reusing the same prime, instead of just doing this:
```suggestion
    p = 32771
    while true
      p = Hecke.next_prime(p)
```